### ISSUE TITLE
slow speed for control recognition

### ISSUE THREAD

**Issue by wangtubs** [pywinauto/pywinauto#329]
Hi, I use pywinauto for iTunes automation on windows. Besides some bugs for control recognition, I also find the speed for control recognition on some window is very slow. I have  to wait over 1 minute to move to next step. I have tried several methods, like exits(), wait(), click_input(), they are all too slow on the specific window. I don't think the window is too complex. 
Do you have any idea how to make it faster?

**Comment by wangtubs** [comment]
to make it more clear, after I fill in some information and click the continue button to submit. using 
`app.iTunes.Static.Pane0.child_window(title="Continue", control_type="Button",visible_only=False).click()`
Then the current window will change to a success window with some text "congratulation"
I then check by using 
`app.iTunes.child_window(title="Congratulations",control_type="Edit").exists(timeout=45)`
the page actually appears in 5s, but I actually wait more than 80s to notice it finally works.
I also tried to find other controls in the menu, it is also very slow to get a response. On other windows, it is not so slow to get response. I can't figure out why on the 'congratulation' window it is so slow. The slowness is not on the iTunes client, it is pywinauto itself.

**Comment by wangtubs** [comment]
here is the info on the success window.
```
Control Identifiers:

Dialog - 'iTunes'    (L407, T134, R1462, B932)
[u'iTunes', u'iTunesDialog', u'Dialog']
child_window(title="iTunes", control_type="Window")
   | 
   | RadioButton - 'Store'    (L1036, T203, R1090, B228)
   | [u'Store', u'StoreRadioButton', 'RadioButton', u'RadioButton1', u'RadioButton0']
   | child_window(title="Store", auto_id="1384670768", control_type="RadioButton")
   | 
   | RadioButton - 'Radio'    (L979, T203, R1035, B228)
   | [u'RadioRadioButton', u'Radio', 'RadioButton2']
   | child_window(title="Radio", auto_id="1384667040", control_type="RadioButton")
   | 
   | RadioButton - 'Browse'    (L913, T203, R978, B228)
   | [u'Browse', u'BrowseRadioButton', 'RadioButton3']
   | child_window(title="Browse", auto_id="1384663312", control_type="RadioButton")
   | 
   | RadioButton - 'For You'    (L844, T203, R912, B228)
   | [u'For You', u'For YouRadioButton', 'RadioButton4']
   | child_window(title="For You", auto_id="1384659584", control_type="RadioButton")
   | 
   | RadioButton - 'Library'    (L780, T203, R843, B228)
   | [u'LibraryRadioButton', u'Library', 'RadioButton5']
   | child_window(title="Library", auto_id="1384655856", control_type="RadioButton")
   | 
   | Static - ''    (L408, T233, R1461, B908)
   | [u'', u'0', u'Static', u'1']
   | child_window(auto_id="1304920864", control_type="Text")
   |    | 
   |    | Pane - ''    (L408, T233, R1461, B908)
   |    | ['Pane', u'2', u'Pane1', u'Pane0']

   |    |    | 
   |    |    | Document - ''    (L408, T233, R1461, B908)
   |    |    | ['Document', u'3']

   |    |    |    | 
   |    |    |    | Custom - ''    (L408, T233, R1461, B908)
   |    |    |    | [u'4', 'Custom8', 'Custom5', 'Custom6', 'Custom7', u'Custom0', u'Custom1', 'Custom2', u'13', u'12', u'18', 'Custom', u'5', u'9']

   |    |    |    |    | 
   |    |    |    |    | Custom - ''    (L1265, T269, R1389, B282)
   |    |    |    |    | [u'4', 'Custom8', 'Custom5', 'Custom6', 'Custom7', u'Custom0', u'Custom1', 'Custom2', u'13', u'12', u'18', 'Custom', u'5', u'9']

   |    |    |    |    |    | 
   |    |    |    |    |    | Edit - 'Secure Connection'    (L1292, T269, R1389, B282)
   |    |    |    |    |    | [u'Edit1', u'Edit0', u'6', 'Edit']
   |    |    |    |    |    | child_window(title="Secure Connection", control_type="Edit")
   |    |    |    |    | 
   |    |    |    |    | Custom - 'Congratulations'    (L480, T253, R1389, B302)
   |    |    |    |    | [u'CongratulationsCustom', 'Custom3', u'Congratulations']
   |    |    |    |    | child_window(title="Congratulations", control_type="Custom")
   |    |    |    |    |    | 
   |    |    |    |    |    | Edit - 'Congratulations'    (L480, T253, R675, B291)
   |    |    |    |    |    | ['Edit2', u'7']
   |    |    |    |    |    | child_window(title="Congratulations", control_type="Edit")
   |    |    |    |    | 
   |    |    |    |    | Custom - 'Your Apple ID is now ready to use with the iTunes Store.'    (L720, T429, R1325, B459)
   |    |    |    |    | [u'Your Apple ID is now ready to use with the iTunes Store.Custom', 'Custom4', u'Your Apple ID is now ready to use with the iTunes Store.']
   |    |    |    |    | child_window(title="Your Apple ID is now ready to use with the iTunes Store.", control_type="Custom")
   |    |    |    |    |    | 
   |    |    |    |    |    | Edit - 'Your Apple ID is now ready to use with the iTunes Store.'    (L720, T433, R1107, B454)
   |    |    |    |    |    | [u'8', 'Edit3']
   |    |    |    |    |    | child_window(title="Your Apple ID is now ready to use with the iTunes Store.", control_type="Edit")
   |    |    |    |    | 
   |    |    |    |    | Custom - ''    (L720, T467, R1325, B481)
   |    |    |    |    | [u'4', 'Custom8', 'Custom5', 'Custom6', 'Custom7', u'Custom0', u'Custom1', 'Custom2', u'13', u'12', u'18', 'Custom', u'5', u'9']

   |    |    |    |    |    | 
   |    |    |    |    |    | Edit - 'You can now shop for your favorite music, movies, TV Shows, and more.'    (L720, T464, R1160, B483)
   |    |    |    |    |    | ['Edit4', u'10']
   |    |    |    |    |    | child_window(title="You can now shop for your favorite music, movies, TV Shows, and more.", control_type="Edit")
   |    |    |    |    | 
   |    |    |    |    | Hyperlink - ''    (L1278, T636, R1389, B659)
   |    |    |    |    | [u'Hyperlink0', u'Hyperlink1', u'11', 'Hyperlink']

   |    |    |    |    |    | 
   |    |    |    |    |    | Button - 'Start Shopping'    (L1280, T636, R1389, B659)
   |    |    |    |    |    | [u'Start ShoppingButton', u'Button1', u'Button0', u'Start Shopping', 'Button']
   |    |    |    |    |    | child_window(title="Start Shopping", control_type="Button")
   |    |    |    |    | 
   |    |    |    |    | Custom - ''    (L480, T867, R1389, B890)
   |    |    |    |    | [u'4', 'Custom8', 'Custom5', 'Custom6', 'Custom7', u'Custom0', u'Custom1', 'Custom2', u'13', u'12', u'18', 'Custom', u'5', u'9']

   |    |    |    |    |    | 
   |    |    |    |    |    | Custom - ''    (L480, T876, R905, B890)
   |    |    |    |    |    | [u'4', 'Custom8', 'Custom5', 'Custom6', 'Custom7', u'Custom0', u'Custom1', 'Custom2', u'13', u'12', u'18', 'Custom', u'5', u'9']

   |    |    |    |    |    |    | 
   |    |    |    |    |    |    | Edit - 'Copyright \xa9 2017 Apple Inc. '    (L480, T876, R628, B889)
   |    |    |    |    |    |    | [u'14', 'Edit5']
   |    |    |    |    |    |    | child_window(title="Copyright \xa9 2017 Apple Inc. ", control_type="Edit")
   |    |    |    |    |    |    | 
   |    |    |    |    |    |    | Hyperlink - 'All rights reserved.'    (L628, T876, R723, B889)
   |    |    |    |    |    |    | ['Hyperlink2', u'All rights reserved.Hyperlink', u'All rights reserved.']
   |    |    |    |    |    |    | child_window(title="All rights reserved.", control_type="Hyperlink")
   |    |    |    |    |    |    |    | 
   |    |    |    |    |    |    |    | Edit - 'All rights reserved.'    (L628, T876, R723, B889)
   |    |    |    |    |    |    |    | ['Edit6', u'15']
   |    |    |    |    |    |    |    | child_window(title="All rights reserved.", control_type="Edit")
   |    |    |    |    |    |    | 
   |    |    |    |    |    |    | Hyperlink - 'Privacy Policy'    (L734, T876, R800, B889)
   |    |    |    |    |    |    | [u'Privacy PolicyHyperlink', 'Hyperlink3', u'Privacy Policy']
   |    |    |    |    |    |    | child_window(title="Privacy Policy", control_type="Hyperlink")
   |    |    |    |    |    |    |    | 
   |    |    |    |    |    |    |    | Edit - 'Privacy Policy'    (L734, T876, R800, B889)
   |    |    |    |    |    |    |    | ['Edit7', u'16']
   |    |    |    |    |    |    |    | child_window(title="Privacy Policy", control_type="Edit")
   |    |    |    |    |    |    | 
   |    |    |    |    |    |    | Hyperlink - 'Terms of Service'    (L816, T876, R897, B889)
   |    |    |    |    |    |    | [u'Terms of ServiceHyperlink', u'Terms of Service', 'Hyperlink4']
   |    |    |    |    |    |    | child_window(title="Terms of Service", control_type="Hyperlink")
   |    |    |    |    |    |    |    | 
   |    |    |    |    |    |    |    | Edit - 'Terms of Service'    (L816, T876, R897, B889)
   |    |    |    |    |    |    |    | ['Edit8', u'17']
   |    |    |    |    |    |    |    | child_window(title="Terms of Service", control_type="Edit")
   |    |    |    |    | 
   |    |    |    |    | Custom - ''    (L408, T873, R1461, B890)
   |    |    |    |    | [u'4', 'Custom8', 'Custom5', 'Custom6', 'Custom7', u'Custom0', u'Custom1', 'Custom2', u'13', u'12', u'18', 'Custom', u'5', u'9']

   | 
   | Edit - 'Search'    (L1207, T140, R1360, B171)
   | ['Edit9', u'19']
   | child_window(title="Search", auto_id="89847440", control_type="Edit")
   | 
   | Pane - 'LCD section'    (L724, T134, R1145, B179)
   | [u'LCD section', u'LCD sectionPane', 'Pane2']
   | child_window(title="LCD section", auto_id="89791584", control_type="Pane")
   | 
   | Button - 'airplay'    (L671, T141, R716, B171)
   | [u'airplay', u'airplayButton', 'Button2']
   | child_window(title="airplay", auto_id="89709696", control_type="Button")
   | 
   | Menu - 'Application'    (L407, T179, R1462, B199)
   | [u'Application', u'ApplicationMenu', u'Menu']
   | child_window(title="Application", auto_id="89161008", control_type="MenuBar")
   |    | 
   |    | MenuItem - 'Song'    (L477, T179, R518, B199)
   |    | [u'MenuItem1', u'MenuItem', u'Song', u'SongMenuItem', u'MenuItem0']
   |    | child_window(title="Song", auto_id="1304907648", control_type="MenuItem")
   |    | 
   |    | MenuItem - 'Help'    (L672, T179, R710, B199)
   |    | [u'Help', u'MenuItem2', u'HelpMenuItem']
   |    | child_window(title="Help", auto_id="90061936", control_type="MenuItem")
   |    | 
   |    | MenuItem - 'Account'    (L615, T179, R672, B199)
   |    | [u'Account', u'MenuItem3', u'AccountMenuItem']
   |    | child_window(title="Account", auto_id="90058272", control_type="MenuItem")
   |    | 
   |    | MenuItem - 'Controls'    (L557, T179, R615, B199)
   |    | [u'Controls', u'ControlsMenuItem', u'MenuItem4']
   |    | child_window(title="Controls", auto_id="90054608", control_type="MenuItem")
   |    | 
   |    | MenuItem - 'View'    (L518, T179, R557, B199)
   |    | [u'ViewMenuItem', u'MenuItem5', u'View']
   |    | child_window(title="View", auto_id="90050944", control_type="MenuItem")
   |    | 
   |    | MenuItem - 'Edit'    (L443, T179, R477, B199)
   |    | [u'Edit10', u'EditMenuItem', u'MenuItem6']
   |    | child_window(title="Edit", auto_id="90047280", control_type="MenuItem")
   |    | 
   |    | MenuItem - 'File'    (L412, T179, R443, B199)
   |    | [u'File', u'MenuItem7', u'FileMenuItem']
   |    | child_window(title="File", auto_id="90043616", control_type="MenuItem")
None
```
I don't think it is too complicated for recognition, because other windows with more text but very fast recognition.

**Comment by airelil** [comment]
Hi @wangtubs, I almost finished my comment when you posted additional details. But I still think this can be useful, so I posting it anyway. 

Yes, "slow automation" can be an issue as it doesn't have one simple solution and there are too many factors affecting it. Just take a look at #256, for instance. You can see that there been many attempts to speed up the things. I would suggest you to keep in mind several "rules of thumb" while writing an automation script on pywinauto:
1. ```win32``` backend usually works faster than ```uia```. On the other hand, sometimes, ```uia``` allows access to controls that are not always accessible from pure ```win32```. You can even go as far as to try automating with two backends at the same time. (Carefully! I didn't try it by myself yet)
2. More detailed criteria in your window specification helps to increase the speed of a control look up. For example, when running ```uia``` backend, the ```control_type``` filter can significantly reduce the search time.
3. Working only through windows specification slows down the things, so you should consider storing and reusing the wrapper of the once resolved control. Simple example, instead of  ```app.Dialog.Button.set_focus(); app.Dialog.Button.click()``` you could write ```wrp = app.Dialog.Button.wrapper_object(); wrp.set_focus(); wrp.click()```. The drawback of this is to make sure that the GUI element you've wrapped in doesn't disappear during your actions.

Back to your question, you may re-think my tip number 2 and try to add an additional filter to your ```child_window``` call. How about looking only for the immediate child, like this: 
```python
app.iTunes.child_window(
    title="Congratulations",
    control_type="Edit",
    top_level_only=True
).exists(timeout=45)
```

**Comment by vasily-v-ryabov** [comment]
In addition to point (3) there is possible to chain some calls on the same control (it will not search for a wrapper twice because `set_focus()` already returned this wrapper):
```python
app.Dialog.Button.set_focus().click()
```

**Comment by wangtubs** [comment]
I tried to use lackey(python version of sikuli) together with pywinauto because lackey is very fast for image recognition. however, there are some compatible issues between pywinauto and lackey
```
    itunes2.child_window(control_type="Edit",found_index=0).type_keys(self.email)
  File "C:\Python27\lib\site-packages\pywinauto\base_wrapper.py", line 832, in type_keys
    turn_off_numlock)
  File "C:\Python27\lib\site-packages\pywinauto\keyboard.py", line 634, in SendKeys
    k.run()
  File "C:\Python27\lib\site-packages\pywinauto\keyboard.py", line 364, in run
    ctypes.sizeof(win32structures.INPUT))
ctypes.ArgumentError: argument 2: <type 'exceptions.TypeError'>: expected LP_INPUT instance instead of pointer to INPUT_Array_2
```
I repeat several times, without import lackey, it works fine. do you know how to fix this bug?
thanks

**Comment by wangtubs** [comment]
in lackey description, he say:
```
Be aware that some Sikuli-script methods actually overwrite Python-native functions, namely type() and input(). Where this is the case, I've remapped the native functions by prefixing them with an underscore. They can be accessed as follows:

from lackey import *

username = _input("Enter your username: ")
```
is it the reason?

**Comment by vasily-v-ryabov** [comment]
Maybe. Redefining Python builtins is very dangerous practice. We renamed all possible re-definitions in pywinauto thanks to static analyzers. I'd recommend you submit a bug against lackey. Who knows how many other libraries can be affected by that.

**Comment by wangtubs** [comment]
hi, so you mean you have fixed this compatible issue from your end? I can upgrade to the latest version?

**Comment by wangtubs** [comment]
hi, thanks. it finally works quite well with lackey. and the speed is much faster than before.

**Comment by vasily-v-ryabov** [comment]
I told about Python built-ins redefinitions that we fixed a long time ago (before 0.6.0). It's not about this particular issue. Maybe you fixed it by changing import order?

**Comment by vasily-v-ryabov** [comment]
The problem seems resolved. I think we can't do anything with UI Automation slowness on some windows.
So the workaround looks like we can close the issue. Please re-open if any.
